### PR TITLE
Added support for writing images using stbi

### PR DESCRIPTION
**Pull request by dpev**
Added unit test for round-trip save/load
Note this pulls the stbi_write implementation into udCore, external projects may have to undefine STB_IMAGE_WRITE_IMPLEMENTATION

### CODE REVIEW

File: Source/udImage.cpp [hunk 1]
```diff
@@ -66,6 +69,90 @@ udResult udImage_LoadFromMemory(udImage **ppImage, const void *pMemory, size_t l
...
+  uint8_t *pRGB = nullptr; // Sadly, stbi has no support for writing 24-bit from a 32-bit source, so we must shrink
+  uint8_t *pSource, *pDest; // For shrinking
+  uint32_t count;
+  stbiWriteContext writeContext;
+
+  UD_ERROR_NULL(pImage, udR_InvalidParameter_);
+  UD_ERROR_NULL(pFilename, udR_InvalidParameter_);
+  UD_ERROR_IF(saveType >= udIST_Max, udR_InvalidParameter_);
+  count = pImage->width * pImage->height;
+  pRGB = (uint8_t *)udAlloc(count * 3);
```
Comment (samsinsane) [inline Source/udImage.cpp:132]: `udAllocType`?

File: Source/udImage.cpp [hunk 2]
```diff
@@ -66,6 +69,90 @@ udResult udImage_LoadFromMemory(udImage **ppImage, const void *pMemory, size_t l
...
+  UD_ERROR_NULL(pRGB, udR_MemoryAllocationFailure);
+  pSource = (uint8_t *)pImage->pImageData;
+  pDest = (uint8_t*)pRGB;
+  while (count--)
+  {
+
+    *pSource++ = pDest[0];
```
Comment (samsinsane) [inline Source/udImage.cpp:139]: I'm guessing you didn't test this before pushing?

**Review by samsinsane — CHANGES_REQUESTED** [review]

File: Source/udImage.cpp (hunk 1 above)
Comment (dpev) [reply Source/udImage.cpp:132]: Oops yeah. It was void in a previous version. cheers

File: Source/udImage.cpp (hunk 2 above)
Comment (dpev) [reply Source/udImage.cpp:139]: Both paths have been tested? This used to do BGR->RGB flipping until I realised that (thankfully!) I didn't end up flipping the bytes inside udImage (though sadly this means I flip them every time we sample). I had branches that did all the flipping, but must have abandoned them thinking we would sooner or later _have_ to return to RGB order. I don't see any problem with how it's done though?

Comment (samsinsane) [reply Source/udImage.cpp:139]: So you intended to assign the garbage memory you allocated above into the source data, and then without setting that garbage memory to anything, save it to disk? The unit test fails on all platforms it runs on too, so something isn't working.

Comment (dpev) [reply Source/udImage.cpp:139]: Hey many garbage memory matters too. But not in this case, pSource is image data and pDest is the RGB buffer? Assigned 134/135 immediately before the loop?

File: Source/udImage.cpp (hunk 1 above)
Comment (dpev) [reply Source/udImage.cpp]: Fixed

File: Source/udImage.cpp (hunk 2 above)
Comment (dpev) [reply Source/udImage.cpp:138]: /facepalm. Obviously I'm already in holiday mode! :) Ok fixing now lol

Comment (samsinsane) [reply Source/udImage.cpp:138]: `pDest[0] = *pSource++;`

Comment (samsinsane) [reply Source/udImage.cpp:138]: It fails on all the build machines, the only place that it matters.

Comment (dpev) [reply Source/udImage.cpp]: Not sure exactly what was going on there but my head is clearly somewhere else today so my memory of tests passing on build machines is probably as wrong as my assignment. Anyway, should be good now :)

**Review by samsinsane — APPROVED** [review]

**Pull request merged**